common/ethereum/proof.go

```diff
@@ -0,0 +1,167 @@
+// This file was adapted from go-ethereum. Here's the go-ethereum license reproduced:
+
+// Copyright 2014 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+// Package trie implements Merkle Patricia Tries.
+
+package ethereum
+
+import (
+	"bytes"
+	"errors"
+	"sync"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/rlp"
+	"github.com/ethereum/go-ethereum/trie"
+)
+
+func NewProof() *Proof {
+	return &Proof{
+		Keys:   make([][]byte, 0),
+		Values: make([][]byte, 0),
+	}
+}
+
+func (m *Proof) Put(key []byte, value []byte) error {
+	for i := 0; i < len(m.Keys); i++ {
+		if bytes.Equal(m.Keys[i], key) {
+			m.Values[i] = value
+			return nil
+		}
+	}
+	m.Keys = append(m.Keys, key)
+	m.Values = append(m.Values, value)
+
+	return nil
+}
+
+func (m *Proof) Delete(key []byte) error {
+	found := false
+	index := -1
+	for i := 0; i < len(m.Keys); i++ {
+		if bytes.Equal(m.Keys[i], key) {
+			found = true
+			index = i
+			continue
+		}
+	}
+	if !found {
+		return errors.New("key not found")
+	}
+	copy(m.Keys[index:len(m.Keys)-1], m.Keys[index+1:])
+	copy(m.Values[index:len(m.Values)-1], m.Values[index+1:])
+	m.Keys = m.Keys[:len(m.Keys)-1]
+	m.Values = m.Values[:len(m.Values)-1]
+
+	return nil
+}
+
+func (m *Proof) Has(key []byte) (bool, error) {
```

Comment:
If this func never throw an error, returning only 'bool' makes caller's life easier.

Reply:
i think this is required interface for using the go-ethereum machinery to generate/verify proofs. 
I remember it was some kind of memdb interface.

Reply:
If interface enforces this, we have no choice then.